keps/sig-node/compute-device-assignment.md

```diff
@@ -0,0 +1,187 @@
+---
+kep-number: 18
+title: Container API and Kubelet extension for device assignment details 
+authors:
+  - "@dashpole" 
+  - "@vikaschoudhary16"
+owning-sig: sig-node
+reviewers:
+  - "@thockin"
+  - "@derekwaynecarr"
+  - "@dchen1107"
+  - "@vishh"
+approvers:
+  - "@sig-node-leads"
+editors:
+  - "@dashpole"
+  - "@vikaschoudhary16"
+creation-date: "2018-07-19"
+last-updated: "2018-07-19"
+status: provisional
+---
+# Container API and Kubelet extension for device assignment details 
+
+Table of Contents
+=================
+* [Abstract](#abstract)
+* [Background](#background)
+* [Objectives](#objectives)
+* [Non Objectives](#non-objectives)
+* [User Journeys](#user-journeys)
+  * [Device Monitoring Agents](#device-monitoring-agents)
+  * [Out-of-tree Experimentation](#out-of-tree-experimentation)
+    * [Network Device Plugins](#network-device-plugins)
+    * [Guaranteed Device Scheduling](#guaranteed-device-scheduling)
+* [Alternatives Considered](#alternatives-considered)
+* [Changes](#changes)
+  * [API Changes](#api-changes)
+  * [Kubelet Changes](#kubelet-changes)
+    * [Pod Admission and Device Allocation](pod-admission-device-allocation)
+    * [Updating ComputeDevices in the pod spec at api server](updating-compute-devices)
+
+## Abstract
+In this document we will discuss the motivation and code changes required for introducing assigned device IDs information in the container API object.
+
+## Background
+The [Pod API Object](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3072) contains the metadata, specification, and status of the smallest unit of computing that can be managed by kubernetes. In order to support Network Devices through the device plugin, it has been proposed (in [Network Devices Support Changes Proposal](https://docs.google.com/document/d/1TpYqgmJQOctm1oexQRo0cGNb9oyk0vozhmhr6myY99M/edit#heading=h.dp2sl8erwljy)) to add Pod UID to the `Allocate` call, and then have the CNI ask the Network Device Plugin for it to do network plumbing. This proposal takes a different approach that does not rely on adding the `PodUID` to the `Allocate` DPAPI call, and does not require explicit coordination between CNI and the Device Plugin.
+
+[Device Monitoring](https://docs.google.com/document/d/1NYnqw-HDQ6Y3L_mk85Q3wkxDtGNWTxpsedsgw4NgWpg/edit?usp=sharing) requires external agents to be able to determine the set of devices in-use by containers and attach pod and container metadata for these devices.
+
+## Objectives
+
+* To enable node monitoring agents to determine the set of devices that are in-use, and to obtain metadata about the container/pod using them.
+* To persist device assignment information outside of the node to simplify and improve kubelet device lifecycle management.
```

Comment:
Can some more clarity be provided on why this is a goal?
Also, can obtaining device information be a best-effort goal - can kubelet re-assign devices to a pod if it cannot recover it's last assignment?
Alternatively, can kubelet use other checkpointing mechanisms like container labels for storing device assignment to containers?

Reply:
@vishh 
> why this is a goal?

There are [several problems due to the use of checkpoint file](https://docs.google.com/document/d/1rYE_jcUjQUqWb-ZRJpG_FLDxOWp5mjOGKGE59HLYGNk/edit?ts=5b6928de#heading=h.47epqgvuoz1s) to rebuild device manager state on node reboot/receate

> can obtaining device information be a best-effort goal?

May be we can if we decide after discussion that this is not a problem worth solving immediately. But i think since we already have enough clarity on problem space from @jiayingz doc, it might be efficient and time-saving at community level to keep it as a main goal. 

>  can kubelet re-assign devices to a pod if it cannot recover it's last assignment?

IMO, it should. However inability to recover  previous assignment is not the only problem. Another for example is how to admit a previously admitted pod, which is requesting a device plugin advertised resource, but is ahead of device plugin pod in serial admission queue? 

> Alternatively, can kubelet use other checkpointing mechanisms like container labels for storing device assignment to containers?

No, i dont think so. To build the device manager state, pods has to go through admission first. Today we use these labels for filling pod status, which is different from this problem.